README.md
[![License](https://img.shields.io/github/license/msolefonte/dublin-pollution-visualizer)](https://github.com/msolefonte/dublin-pollution-visualizer/blob/main/LICENSE)
[![contributions welcome](https://img.shields.io/badge/contributions-welcome-brightgreen.svg)](https://github.com/msolefonte/dublin-pollution-visualizer/CONTRIBUTING.md)

# Dublin NO<sub>2</sub> Pollution Visualizer

Open Source application that allows visualization real-time data about traffic and Nitrogen Dioxide pollution around Dublin, Ireland. To be used together with Dublin Pollution Piper.

## Usage

Nothing to install, nothing to build, nothing to worry about! Just a static website.

## Contributing

Feel free to add issues or to create pull requests. Help is always welcome.

## Versioning

[SemVer](http://semver.org/) is used for versioning. For the changelog, see [CHANGELOG.md](CHANGELOG.md).

## Authors

* **Marc Solé Fonte** - *Initial work* - [msolefonte](https://github.com/msolefonte)

## License

This project is licensed under the Apache License, Version 3.0 - see the [LICENSE](LICENSE) file for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.2] - 20/11/2021

- Initial release


js/base.js
"use strict";

function _createForOfIteratorHelper(o, allowArrayLike) { var it; if (typeof Symbol === "undefined" || o[Symbol.iterator] == null) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = o[Symbol.iterator](); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

document.addEventListener('DOMContentLoaded', function () {
  'use strict';

  feather.replace();

  (function () {
    var sidebar = document.querySelector('.sidebar'),
        catSubMenu = document.querySelector('.cat-sub-menu'),
        sidebarBtns = document.querySelectorAll('.sidebar-toggle');

    var _iterator = _createForOfIteratorHelper(sidebarBtns),
        _step;

    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var sidebarBtn = _step.value;

        if (sidebarBtn && catSubMenu && sidebarBtn) {
          sidebarBtn.addEventListener('click', function () {
            var _iterator2 = _createForOfIteratorHelper(sidebarBtns),
                _step2;

            try {
              for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
                var sdbrBtn = _step2.value;
                sdbrBtn.classList.toggle('rotated');
              }
            } catch (err) {
              _iterator2.e(err);
            } finally {
              _iterator2.f();
            }

            sidebar.classList.toggle('hidden');
            catSubMenu.classList.remove('visible');
          });
        }
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
  })();

  (function () {
    var showCatBtns = document.querySelectorAll('.show-cat-btn');

    if (showCatBtns) {
      showCatBtns.forEach(function (showCatBtn) {
        var catSubMenu = showCatBtn.nextElementSibling;
        showCatBtn.addEventListener('click', function (e) {
          e.preventDefault();
          catSubMenu.classList.toggle('visible');
          var catBtnToRotate = document.querySelector('.category__btn');
          catBtnToRotate.classList.toggle('rotated');
        });
      });
    }
  })();

  (function () {
    var showMenu = document.querySelector('.lang-switcher');
    var langMenu = document.querySelector('.lang-menu');
    var layer = document.querySelector('.layer');

    if (showMenu) {
      showMenu.addEventListener('click', function () {
        langMenu.classList.add('active');
        layer.classList.add('active');
      });

      if (layer) {
        layer.addEventListener('click', function (e) {
          if (langMenu.classList.contains('active')) {
            langMenu.classList.remove('active');
            layer.classList.remove('active');
          }
        });
      }
    }
  })();

  (function () {
    var userDdBtnList = document.querySelectorAll('.dropdown-btn');
    var userDdList = document.querySelectorAll('.users-item-dropdown');
    var layer = document.querySelector('.layer');

    if (userDdList && userDdBtnList) {
      var _iterator3 = _createForOfIteratorHelper(userDdBtnList),
          _step3;

      try {
        for (_iterator3.s(); !(_step3 = _iterator3.n()).done;) {
          var userDdBtn = _step3.value;
          userDdBtn.addEventListener('click', function (e) {
            layer.classList.add('active');

            var _iterator4 = _createForOfIteratorHelper(userDdList),
                _step4;

            try {
              for (_iterator4.s(); !(_step4 = _iterator4.n()).done;) {
                var userDd = _step4.value;

                if (e.currentTarget.nextElementSibling == userDd) {
                  if (userDd.classList.contains('active')) {
                    userDd.classList.remove('active');
                  } else {
                    userDd.classList.add('active');
                  }
                } else {
                  userDd.classList.remove('active');
                }
              }
            } catch (err) {
              _iterator4.e(err);
            } finally {
              _iterator4.f();
            }
          });
        }
      } catch (err) {
        _iterator3.e(err);
      } finally {
        _iterator3.f();
      }
    }

    if (layer) {
      layer.addEventListener('click', function (e) {
        var _iterator5 = _createForOfIteratorHelper(userDdList),
            _step5;

        try {
          for (_iterator5.s(); !(_step5 = _iterator5.n()).done;) {
            var userDd = _step5.value;

            if (userDd.classList.contains('active')) {
              userDd.classList.remove('active');
              layer.classList.remove('active');
            }
          }
        } catch (err) {
          _iterator5.e(err);
        } finally {
          _iterator5.f();
        }
      });
    }
  })();

  (function () {
    Chart.defaults.backgroundColor = '#000';
    var darkMode = localStorage.getItem('darkMode');
    var darkModeToggle = document.querySelector('.theme-switcher');

    var enableDarkMode = function enableDarkMode() {
      document.body.classList.add('darkmode');
      localStorage.setItem('darkMode', 'enabled');
    };

    var disableDarkMode = function disableDarkMode() {
      document.body.classList.remove('darkmode');
      localStorage.setItem('darkMode', null);
    };

    if (darkMode === 'enabled') {
      enableDarkMode();
    }

    if (darkModeToggle) {
      darkModeToggle.addEventListener('click', function () {
        darkMode = localStorage.getItem('darkMode');

        if (darkMode !== 'enabled') {
          enableDarkMode();
        } else {
          disableDarkMode();
        }

        addData();
      });
    }
  })();

  (function () {
    var checkAll = document.querySelector('.check-all');
    var checkers = document.querySelectorAll('.check');

    if (checkAll && checkers) {
      checkAll.addEventListener('change', function (e) {
        var _iterator6 = _createForOfIteratorHelper(checkers),
            _step6;

        try {
          for (_iterator6.s(); !(_step6 = _iterator6.n()).done;) {
            var checker = _step6.value;

            if (checkAll.checked) {
              checker.checked = true;
              checker.parentElement.parentElement.parentElement.classList.add('active');
            } else {
              checker.checked = false;
              checker.parentElement.parentElement.parentElement.classList.remove('active');
            }
          }
        } catch (err) {
          _iterator6.e(err);
        } finally {
          _iterator6.f();
        }
      });

      var _iterator7 = _createForOfIteratorHelper(checkers),
          _step7;

      try {
        var _loop = function _loop() {
          var checker = _step7.value;
          checker.addEventListener('change', function (e) {
            checker.parentElement.parentElement.parentElement.classList.toggle('active');

            if (!checker.checked) {
              checkAll.checked = false;
            }

            var totalCheckbox = document.querySelectorAll('.users-table .check');
            var totalChecked = document.querySelectorAll('.users-table .check:checked');

            if (totalCheckbox && totalChecked) {
              if (totalCheckbox.length == totalChecked.length) {
                checkAll.checked = true;
              } else {
                checkAll.checked = false;
              }
            }
          });
        };

        for (_iterator7.s(); !(_step7 = _iterator7.n()).done;) {
          _loop();
        }
      } catch (err) {
        _iterator7.e(err);
      } finally {
        _iterator7.f();
      }
    }
  })();

  (function () {
    var checkAll = document.querySelector('.check-all');
    var checkers = document.querySelectorAll('.check');
    var checkedSum = document.querySelector('.checked-sum');

    if (checkedSum && checkAll && checkers) {
      checkAll.addEventListener('change', function (e) {
        var totalChecked = document.querySelectorAll('.users-table .check:checked');
        checkedSum.textContent = totalChecked.length;
      });

      var _iterator8 = _createForOfIteratorHelper(checkers),
          _step8;

      try {
        for (_iterator8.s(); !(_step8 = _iterator8.n()).done;) {
          var checker = _step8.value;
          checker.addEventListener('change', function (e) {
            var totalChecked = document.querySelectorAll('.users-table .check:checked');
            checkedSum.textContent = totalChecked.length;
          });
        }
      } catch (err) {
        _iterator8.e(err);
      } finally {
        _iterator8.f();
      }
    }
  })();

  var charts = {};
  var gridLine;
  var titleColor;

  (function () {
    /* Add gradient to chart */
    var width, height, gradient;

    function getGradient(ctx, chartArea) {
      var chartWidth = chartArea.right - chartArea.left;
      var chartHeight = chartArea.bottom - chartArea.top;

      if (gradient === null || width !== chartWidth || height !== chartHeight) {
        width = chartWidth;
        height = chartHeight;
        gradient = ctx.createLinearGradient(0, chartArea.bottom, 0, chartArea.top);
        gradient.addColorStop(0, 'rgba(255, 255, 255, 0)');
        gradient.addColorStop(1, 'rgba(255, 255, 255, 0.4)');
      }

      return gradient;
    }
    /* Visitors chart */

    var ctx = document.getElementById('myChart');

    if (ctx) {
      var myCanvas = ctx.getContext('2d');
      var myChart = new Chart(myCanvas, {
        type: 'line',
        data: {
          labels: ['Dec', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'],
          datasets: [{
            label: 'Last 6 months',
            data: [35, 27, 40, 15, 30, 25, 45],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
          }, {
            label: 'Previous',
            data: [20, 36, 16, 45, 29, 32, 10],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(75, 222, 151, 1)'],
            borderColor: ['rgba(75, 222, 151, 1)'],
            borderWidth: 2
          }]
        },
        options: {
          scales: {
            y: {
              min: 0,
              max: 100,
              ticks: {
                stepSize: 25
              },
              grid: {
                display: false
              }
            },
            x: {
              grid: {
                color: gridLine
              }
            }
          },
          elements: {
            point: {
              radius: 2
            }
          },
          plugins: {
            legend: {
              position: 'top',
              align: 'end',
              labels: {
                boxWidth: 8,
                boxHeight: 8,
                usePointStyle: true,
                font: {
                  size: 12,
                  weight: '500'
                }
              }
            },
            title: {
              display: true,
              text: ['Visitor statistics', 'Nov - July'],
              align: 'start',
              color: '#171717',
              font: {
                size: 16,
                family: 'Inter',
                weight: '600',
                lineHeight: 1.4
              }
            }
          },
          tooltips: {
            mode: 'index',
            intersect: false
          },
          hover: {
            mode: 'nearest',
            intersect: true
          }
        }
      });
      charts.visitors = myChart;
    }
    /* Customers chart */

    var customersChart = document.getElementById('customersChart');

    if (customersChart) {
      var customersChartCanvas = customersChart.getContext('2d');
      var myCustomersChart = new Chart(customersChartCanvas, {
        type: 'line',
        data: {
          labels: ['Dec', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun'],
          datasets: [{
            label: '+958',
            data: [90, 10, 80, 20, 70, 30, 50],
            tension: 0.4,
            backgroundColor: function backgroundColor(context) {
              var chart = context.chart;
              var ctx = chart.ctx,
                  chartArea = chart.chartArea;

              if (!chartArea) {
                // This case happens on initial chart load
                return null;
              }

              return getGradient(ctx, chartArea);
            },
            borderColor: ['#fff'],
            borderWidth: 2,
            fill: true
          }]
        },
        options: {
          scales: {
            y: {
              display: false
            },
            x: {
              display: false
            }
          },
          elements: {
            point: {
              radius: 1
            }
          },
          plugins: {
            legend: {
              position: 'top',
              align: 'end',
              labels: {
                color: '#fff',
                size: 18,
                fontStyle: 800,
                boxWidth: 0
              }
            },
            title: {
              display: true,
              text: ['New Customers', '28 Daily Avg.'],
              align: 'start',
              color: '#fff',
              font: {
                size: 16,
                family: 'Inter',
                weight: '600',
                lineHeight: 1.4
              },
              padding: {
                top: 20
              }
            }
          },
          maintainAspectRatio: false
        }
      });
      customersChart.customers = myCustomersChart;
    }
  })();
  /* Change data of all charts */


  function addData() {
    var darkMode = localStorage.getItem('darkMode');

    if (darkMode === 'enabled') {
      gridLine = '#37374F';
      titleColor = '#EFF0F6';
    } else {
      gridLine = '#EEEEEE';
      titleColor = '#171717';
    }

    if (charts.hasOwnProperty('visitors')) {
      charts.visitors.options.scales.x.grid.color = gridLine;
      charts.visitors.options.plugins.title.color = titleColor;
      charts.visitors.options.scales.y.ticks.color = titleColor;
      charts.visitors.options.scales.x.ticks.color = titleColor;
      charts.visitors.update();
    }
  }

  addData();
});


js/gatherer.js
"use strict";

// var apiUrl = 'http://localhost:8080/v1/';
var apiUrl = 'http://13.80.18.173:8080/v1/';

function updateTotalEmissions(thisMonth, previousMonth) {
    if (document.getElementById("dashboard-emissions-monthly")) {
        document.getElementById("dashboard-emissions-monthly").innerHTML = Math.round(thisMonth) / 1000 + ' kg';

        const evolution = Math.round(thisMonth / previousMonth * 100) / 100;
        document.getElementById("dashboard-emissions-evolution").innerHTML =
            '<span  class="stat-cards-info__profit danger ">' +
            '<i data-feather="trending-up" aria-hidden="true"></i>+' + evolution + '%' +
            '</span>Last month';
    }
}

function updateLastEmissionValue(area, value) {
    if (document.getElementById("area-" + area + "-last-emission-value")) {
        if (typeof value !== "undefined") {
            var flag = 'danger';

            if (value < 25) {
                flag = 'success';
            } else if (value < 50) {
                flag = 'warning';
            }

            document.getElementById("area-" + area + "-last-emission-value").innerHTML =
                '<span  class="stat-cards-info__profit ' + flag + '">' +
                '<i data-feather="cloud" aria-hidden="true"></i>' + value + ' μg/m³' +
                '</span>';
        } else {
            document.getElementById("area-" + area + "-last-emission-value").innerHTML =
                '<span  class="stat-cards-info__profit danger ">' +
                '<i data-feather="warning" aria-hidden="true"></i> Sensor not available' +
                '</span>';
        }
    }
}



function updateColor(chart, firstCall) {
    const darkMode = localStorage.getItem('darkMode');

    let titleColor;

    if (darkMode === 'enabled') {
        titleColor = '#EFF0F6';
    } else {
        titleColor = '#171717';
    }

    chart.options.plugins.title.color = titleColor;
    chart.update();
}

function updateEmissionsChart(labels, normalized_emissions, emissions, traffic) {
    const options = {
            spanGaps: 1000 * 60 * 60 * 24 * 2, // 2 days
            responsive: true,
            scales: {
                x: [{
                    type: 'time',
                    title: {
                        display: true,
                        text: 'Date'
                    },
                    ticks: {
                        autoSkip: false,
                        maxRotation: 0,
                        major: {
                            enabled: true
                        },
                        font: function (context) {
                            if (context.tick && context.tick.major) {
                                return {
                                    weight: 'bold',
                                };
                            }
                        }
                    }
                }]
            },
            elements: {
                point: {
                    radius: 2
                }
            },
            plugins: {
                legend: {
                    position: 'top',
                    align: 'end',
                    labels: {
                        boxWidth: 8,
                        boxHeight: 8,
                        usePointStyle: true,
                        font: {
                            size: 12,
                            weight: '500'
                        }
                    }
                },
                title: {
                    display: true,
                    text: ['Emissions this week (μg/m³)'],
                    align: 'start',
                    color: '#171717',
                    font: {
                        size: 16,
                        family: 'Inter',
                        weight: '600',
                        lineHeight: 1.4
                    }
                }
            },
            tooltips: {
                mode: 'index',
                intersect: false
            },
            hover: {
                mode: 'nearest',
                intersect: true
            }
        };

    const data = {
        labels: labels,
        datasets: [{
            label: 'South Canal',
            data: normalized_emissions[0],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        }, {
            label: 'The Liberties',
            data: normalized_emissions[1],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgb(227,234,130, 1)'],
            borderColor: ['rgba(227,234,130, 1)'],
            borderWidth: 2
        }, {
            label: 'Dublin Heuston',
            data: normalized_emissions[2],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 146, 234, 1)'],
            borderColor: ['rgba(95, 146, 234, 1)'],
            borderWidth: 2
        }, {
            label: 'Tech. University',
            data: normalized_emissions[3],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgb(234,35,53, 1)'],
            borderColor: ['rgba(234,35,53, 1)'],
            borderWidth: 2
        }, {
            label: 'North City Center',
            data: normalized_emissions[4],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 134, 1)'],
            borderColor: ['rgba(95, 46, 134, 1)'],
            borderWidth: 2
        }, {
            label: 'City Port',
            data: normalized_emissions[5],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(75, 222, 151, 1)'],
            borderColor: ['rgba(75, 222, 151, 1)'],
            borderWidth: 2
        }]
    };

    const data1 = {
        labels: labels,
        datasets: [{
            label: 'Emissions',
            data: emissions[0],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        },{
            label: 'Avg. Travel Time (s)',
            data: traffic[0],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(245, 230, 40, 1)'],
            borderColor: ['rgba(245, 230, 40, 1)'],
            borderWidth: 2
        }]
    };

    const data2 = {
        labels: labels,
        datasets: [{
            label: 'Emissions',
            data: emissions[1],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        },{
            label: 'Avg. Travel Time (s)',
            data: traffic[1],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(245, 230, 40, 1)'],
            borderColor: ['rgba(245, 230, 40, 1)'],
            borderWidth: 2,
            spanGaps: false
        }]
    };

    const data3 = {
        labels: labels,
        datasets: [{
            label: 'Emissions',
            data: emissions[2],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        },{
            label: 'Avg. Travel Time (s)',
            data: traffic[2],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(245, 230, 40, 1)'],
            borderColor: ['rgba(245, 230, 40, 1)'],
            borderWidth: 2
        }]
    };

    const data4 = {
        labels: labels,
        datasets: [{
            label: 'Emissions',
            data: emissions[3],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        },{
            label: 'Avg. Travel Time (s)',
            data: traffic[3],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(245, 230, 40, 1)'],
            borderColor: ['rgba(245, 230, 40, 1)'],
            borderWidth: 2
        }]
    };

    const data5 = {
        labels: labels,
        datasets: [{
            label: 'Emissions',
            data: emissions[4],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        },{
            label: 'Avg. Travel Time (s)',
            data: traffic[4],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(245, 230, 40, 1)'],
            borderColor: ['rgba(245, 230, 40, 1)'],
            borderWidth: 2
        }]
    };

    const data6 = {
        labels: labels,
        datasets: [{
            label: 'Emissions',
            data: emissions[5],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(95, 46, 234, 1)'],
            borderColor: ['rgba(95, 46, 234, 1)'],
            borderWidth: 2
        },{
            label: 'Avg. Travel Time (s)',
            data: traffic[5],
            cubicInterpolationMode: 'monotone',
            tension: 0.4,
            backgroundColor: ['rgba(245, 230, 40, 1)'],
            borderColor: ['rgba(245, 230, 40, 1)'],
            borderWidth: 2
        }]
    };

    const emissionsChartCtx = document.getElementById('dashboard-emissions-chart');
    if (emissionsChartCtx) {
        const emissionsCanvas = emissionsChartCtx.getContext('2d');
        const emissionsChart = new Chart(emissionsCanvas, {
            type: "line",
            options: options,
            data: data,
        });

        updateColor(emissionsChart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissionsChart);
        };
    }

    const emissions1ChartCtx = document.getElementById('dashboard-emissions-1-chart');
    if (emissions1ChartCtx) {
        const emissions1Canvas = emissions1ChartCtx.getContext('2d');
        const emissions1Chart = new Chart(emissions1Canvas, {
            type: "line",
            options: options,
            data: data1,
        });

        updateColor(emissions1Chart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissions1Chart);
        };
    }

    const emissions2ChartCtx = document.getElementById('dashboard-emissions-2-chart');
    if (emissions2ChartCtx) {
        const emissions2Canvas = emissions2ChartCtx.getContext('2d');
        const emissions2Chart = new Chart(emissions2Canvas, {
            type: "line",
            options: options,
            data: data2,
        });

        updateColor(emissions2Chart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissions2Chart);
        };
    }

    const emissions3ChartCtx = document.getElementById('dashboard-emissions-3-chart');
    if (emissions3ChartCtx) {
        const emissions3Canvas = emissions3ChartCtx.getContext('2d');
        const emissions3Chart = new Chart(emissions3Canvas, {
            type: "line",
            options: options,
            data: data3,
        });

        updateColor(emissions3Chart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissions3Chart);
        };
    }

    const emissions4ChartCtx = document.getElementById('dashboard-emissions-4-chart');
    if (emissions4ChartCtx) {
        const emissions4Canvas = emissions4ChartCtx.getContext('2d');
        const emissions4Chart = new Chart(emissions4Canvas, {
            type: "line",
            options: options,
            data: data4,
        });

        updateColor(emissions4Chart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissions4Chart);
        };
    }

    const emissions5ChartCtx = document.getElementById('dashboard-emissions-5-chart');
    if (emissions5ChartCtx) {
        const emissions5Canvas = emissions5ChartCtx.getContext('2d');
        const emissions5Chart = new Chart(emissions5Canvas, {
            type: "line",
            options: options,
            data: data5,
        });

        updateColor(emissions5Chart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissions5Chart);
        };
    }

    const emissions6ChartCtx = document.getElementById('dashboard-emissions-6-chart');
    if (emissions6ChartCtx) {
        const emissions6Canvas = emissions6ChartCtx.getContext('2d');
        const emissions6Chart = new Chart(emissions6Canvas, {
            type: "line",
            options: options,
            data: data6,
        });

        updateColor(emissions6Chart);

        document.getElementById('theme-switcher').onclick = function () {
            updateColor(emissions6Chart);
        };
    }
}

async function updateAllEmissionValues() {
    const one_day_from_now = new Date(new Date().getTime() - 24 * 60 * 60 * 1000);
    const one_week_from_then = new Date(one_day_from_now.getTime() - 7 * 24 * 60 * 60 * 1000);
    const one_month_from_then = new Date(one_day_from_now.getTime() - 31 * 24 * 60 * 60 * 1000);
    const two_months_from_then = new Date(one_day_from_now.getTime() - 62 * 24 * 60 * 60 * 1000);

    const chart_emissions = [];
    const chart_emissions_normalized = [];
    const chart_emissions_labels = {};
    const chart_traffic = [];
    let this_month_total_emissions = 0;
    let previous_month_total_emissions = 0;

    for (let area = 1; area <= 6; area++) {
        const emissionsResponse = await fetch(apiUrl + 'emissions/' + area);
        const emissionsData = await emissionsResponse.json();

        const trafficResponse = await fetch(apiUrl + 'traffic/' + area);
        const trafficData = await trafficResponse.json();

        if (emissionsData.length > 0) {
            updateLastEmissionValue(area, emissionsData[0].no2_level);
        } else {
            updateLastEmissionValue(area, undefined);
        }

        const chart_emissions_local = [];
        const chart_emissions_normalized_local = [];
        const chart_traffic_local = [];

        for (let i = 0; i < emissionsData.length; i++) {
            const entry_date = new Date(emissionsData[i].timestamp);

            if (entry_date > one_week_from_then) {;
                chart_emissions_local.push(
                    {x: entry_date, y: emissionsData[i].no2_level}
                );
                chart_emissions_normalized_local.push(
                    {x: entry_date.toISOString().substring(5, 16), y: emissionsData[i].no2_level}
                );
                this_month_total_emissions += emissionsData[i].no2_level;
                chart_emissions_labels[entry_date.toISOString().substring(5, 16)] = 1;
            } else if (entry_date > one_month_from_then) {
                this_month_total_emissions += emissionsData[i].no2_level;
            } else if (entry_date > two_months_from_then) {
                previous_month_total_emissions += emissionsData[i].no2_level;
            } else {
                break;
            }
        }

        const averaged_traffic = {};
        for (let i = 0; i < trafficData.length; i++) {
            const entry_date = new Date(trafficData[i].timestamp);
            entry_date.setMinutes(0);
            entry_date.setSeconds(0);

            if (entry_date in averaged_traffic) {
                averaged_traffic[entry_date][0] += trafficData[i].current_travel_time;
                averaged_traffic[entry_date][1] += 1
            } else {
                averaged_traffic[entry_date] = [trafficData[i].current_travel_time, 1]
            }
        }

        for (let i = 0; i < Object.keys(averaged_traffic).length; i++) {
            const key = Object.keys(averaged_traffic)[i];
            const entry_date = new Date(key);

            if (entry_date > one_week_from_then) {
                chart_traffic_local.push(
                    {
                        x: new Date(key),
                        y: averaged_traffic[key][0] / averaged_traffic[key][1]
                    }
                );
            } else {
                break;
            }
        }

        chart_emissions.push(chart_emissions_local);
        chart_traffic.push(chart_traffic_local);
        chart_emissions_normalized.push(chart_emissions_normalized_local);
    }

    updateTotalEmissions(this_month_total_emissions, previous_month_total_emissions);
    updateEmissionsChart(Object.keys(chart_emissions_labels).sort(), chart_emissions_normalized, chart_emissions, chart_traffic);
}

async function updateDataPointsAvailable() {
    if (document.getElementById("dashboard-available-data-points")) {
        const response = await fetch(apiUrl + 'data-points');
        const data = await response.json();

        document.getElementById("dashboard-available-data-points").innerHTML = data;
    }
}


async function updateLastUpdate() {
    if (document.getElementById("dashboard-last-update")) {
        const response = await fetch(apiUrl + 'last-update');
        const data = await response.json();

        document.getElementById("dashboard-last-update").innerHTML =
            't ' + new Date(data).toTimeString().substring(0, 8);
    }
}

updateAllEmissionValues();
updateDataPointsAvailable();
updateLastUpdate();
